src/Common/ARCFileCache.cpp

```diff
@@ -0,0 +1,925 @@
+            "Nothing to reduce, file segment fully downloaded, key: {}, offset: {}",
+            keyToStr(key),
+            offset);
+
+    cell->file_segment = std::make_shared<FileSegment>(offset, downloaded_size, key, this, FileSegment::State::DOWNLOADED);
+}
+
+bool ARCFileCache::isLastFileSegmentHolder(
+    const Key & key, size_t offset, std::lock_guard<std::mutex> & cache_lock, std::lock_guard<std::mutex> & /* segment_lock */)
+{
+    auto * cell = getCell(key, offset, cache_lock);
+
+    if (!cell)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "No cell found for key: {}, offset: {}", keyToStr(key), offset);
+
+    /// The caller of this method is last file segment holder if use count is 2 (the second pointer is cache itself)
+    return cell->file_segment.use_count() == 2;
+}
+
+bool ARCFileCache::canMoveCellToHighQueue(const FileSegmentCell & cell) const
+{
+    return cell.hit_count >= move_threshold;
+}
+
+bool ARCFileCache::tryMoveLowToHigh(const FileSegmentCell & cell, std::lock_guard<std::mutex> & cache_lock)
+{
+    int increase_size = 0;
+    if (high_queue.max_space_bytes() < max_high_space_size)
+    {
+        increase_size = std::min(cell.size(), max_high_space_size - high_queue.max_space_bytes());
+        high_queue.increase_max_space_bytes(increase_size);
+        low_queue.increase_max_space_bytes(0 - increase_size);
+    }
+
+    if (tryReserve(high_queue, cell.file_segment->key(), cell.file_segment->offset(), cell.size(), cache_lock))
+    {
+        cell.is_low = false;
+        high_queue.queue().splice(high_queue.queue().end(), low_queue.queue(), *cell.queue_iterator);
+        low_queue.increase_space_bytes(0 - cell.size());
+        return true;
+    }
+    else
+    {
+        /// rollback
+        high_queue.increase_max_space_bytes(0 - increase_size);
+        low_queue.increase_max_space_bytes(increase_size);
+        return false;
+    }
+}
+
+ARCFileCache::FileSegmentCell::FileSegmentCell(FileSegmentPtr file_segment_, LRUQueue & queue_) : file_segment(file_segment_)
+{
+    /**
+     * Cell can be created with either DOWNLOADED or EMPTY file segment's state.
+     * File segment acquires DOWNLOADING state and creates LRUQueue iterator on first
+     * successful getOrSetDownaloder call.
+     */
+
+    switch (file_segment->download_state)
+    {
+        case FileSegment::State::DOWNLOADED: {
+            queue_iterator = queue_.insert(queue_.end(), getKeyAndOffset());
+            break;
+        }
+        case FileSegment::State::EMPTY:
+        case FileSegment::State::DOWNLOADING: {
+            break;
+        }
+        default:
+            throw Exception(
+                ErrorCodes::REMOTE_FS_OBJECT_CACHE_ERROR,
+                "Can create cell with either EMPTY, DOWNLOADED, DOWNLOADING state, got: {}",
+                FileSegment::stateToString(file_segment->download_state));
+    }
+}
+
+String ARCFileCache::dumpStructure(const Key & key_)
```

Comment:
Let's add some basic unit tests which check ARC cache policy correctness. See `src/Common/tests/gtest_lru_file_cache.cpp`

How to test locally:
```
cd build
ninja unit_tests_dbms
./src/unit_tests_dbms --gtest_filter='LRUFileCache*' --gtest_brief=0 --gtest_throw_on_failure=1
```

Reply:
> Let's add some basic unit tests which check ARC cache policy correctness. See `src/Common/tests/gtest_lru_file_cache.cpp`
> 
> How to test locally:
> 
> ```
> cd build
> ninja unit_tests_dbms
> ./src/unit_tests_dbms --gtest_filter='LRUFileCache*' --gtest_brief=0 --gtest_throw_on_failure=1
> ```

okay